### ISSUE TITLE
timeout does not work with https connections and JRuby

### ISSUE THREAD

**Issue by dgolombek** [nahi/httpclient#341]
This is very similar to #206, but for https connections. Taking a page from @ocher's book, sample code:

```
require 'rubygems'
require 'socket'
require 'thread'
require 'httpclient'
require 'test/unit'

class JRubyHTTPClientReceiveTimeoutBugTest < Test::Unit::TestCase
  TIMEOUT = 1 # seconds

  def test_should_raise_exception_when_receiving_data_takes_too_long
    httpclient = HTTPClient.new
    httpclient.connect_timeout = TIMEOUT
    httpclient.receive_timeout = TIMEOUT
    httpclient.send_timeout = TIMEOUT
    httpclient.ssl_config.timeout = TIMEOUT

    begin
      puts "Connecting"
      start_time = Time.now
      url = "https://www.google.com:81"
      httpclient.get(url)
    rescue HTTPClient::ReceiveTimeoutError, HTTPClient::ConnectTimeoutError => e
      end_time = Time.now
      total_time = end_time - start_time
      assert total_time < TIMEOUT + 1, "Receiving data took too much time: #{total_time}"
    end
  end
end
```

Running with MRI 2.3.0, that works just fine. With JRuby 9.1.5.0, it takes ~30 seconds to timeout. 

UPDATE: This is a HTTPClient issue, not a JRuby bug. See proposed fix in #342.

OLD:
If I understand correctly, I think that at https://github.com/nahi/httpclient/blob/master/lib/httpclient/jruby_ssl_socket.rb#L444 we should only create the Socket if a proxy is specified. If no proxy is present, at https://github.com/nahi/httpclient/blob/master/lib/httpclient/jruby_ssl_socket.rb#L500 we should be doing

```
socket = factory.createSocket(); 
socket_addr = InetSocketAddress.new(dest.site.host, dest.site.port)
if session.ssl_config.timeout
    socket.connect(socket_addr, session.ssl_config.timeout / 1000); 
else
    socket.connect(socket_addr); 
end
socket.setSoTimeout(session.ssl_config.timeout / 1000) if session.ssl_config.timeout
```

That's pretty rough, but I think it's why no timeout is happening. It does point out that ssl_config only has a single timeout instead of both a connect_timeout and send/receive timeouts. It also makes me wonder why there is a separate SSL timeout at all, instead of automatically sharing the default timeouts.

Note: In my test, I'm using a known host with firewall that's dropping SYN  packets here to simulate a really bad case. If I change the test from #206 to use https, it times out after 3 seconds -- better, although still not perfect. But I did run into this bug in the real world, with a host that was failing in a really bad way.

**Comment by dgolombek** [comment]
In all the timeout logic above, it should clearly be * 1000, not / 1000.

**Comment by dgolombek** [comment]
Reading through the code more, it appears that the SSLConfig.timeout is only used for JRuby, not for MRI. Given that, I propose that ssl_config should no longer have it's own timeout value, and instead only act as a proxy for the connect_timeout and read_timeout values from the HTTPClient object itself. If that makes sense, I'll create a separate issue and apply that fix on top of this one.

**Comment by nahi** [comment]
@dgolombek Thanks for the investigation and patch. I'm still reading those slowly (sorry.)
Re: SSLConfig#timeout, it's the timeout of SSLSession object, not connection.
And unlike the httpclient code when #206 was reported, now we use Java socket directly so httpclient need to care about `connect_timeout` by itself, perhaps by specifying so_timeout as you did in #342.

**Comment by dgolombek** [comment]
Ah, thanks for clarification on the timeout -- I should've traced that through more. My change is incorrect as is, with that clarification, since it's using ssl_config.timeout rather than httpclient.connect_timeout. And setSoTimeout should use httpclient.receive_timeout. I'll update the patch with those changes.

**Comment by dgolombek** [comment]
For an example of setting this connect timeout from elsewhere, see http://stackoverflow.com/questions/5715751/ssl-socket-connect-timeout

**Comment by dgolombek** [comment]
I updated the patch to use the correct timeout values.

**Comment by dgolombek** [comment]
I realized one other advantage of this approach -- it moves the socket creation to within the existing begin/rescue block that translates Java Exceptions into Ruby exceptions. Thus no more Java::JavaNet::NoRouteToHostException and Java::JavaNet::ConnectException being raised. However, it still won't handle java.net.UnknownHostException correctly -- I'll fix that separately, along with some missing exception handling in HTTPClient::JavaSocketWrap.fill.

**Comment by dgolombek** [comment]
I created #345 to address the last comment and added a PR.

**Comment by nahi** [comment]
I merged #346, updated version of #345 by @dgolombek to solve this problem. Thank you very much for the effort.

**Comment by dgolombek** [comment]
Thank you @nahi!
